File: LRSMessageToolBar/Classes/Emoji/LRSMemePackagesView.m


#import "LRSMemePackagesView.h"
#import "LRSMemePackageControlView.h"
#import "LRSMemePackageConfigure.h"
#import "LRSMessageToolBarHelper.h"
#import <MASonry/MASonry.h>

@interface LRSMemePackagesView () <UIScrollViewDelegate>
@property (nonatomic, strong) UIButton *sendEmojiButton;
@property (nonatomic, strong) UIScrollView *mainScrollView;
@property (nonatomic, strong) UIScrollView *barScrollView;
@property (nonatomic, strong) NSArray<LRSMemePackageConfigure *> *configures;
@end


@implementation LRSMemePackagesView

- (instancetype)initWithFrame:(CGRect)frame configures:(NSArray<LRSMemePackageConfigure *> *)configures {
    if (self = [super initWithFrame:frame]) {
        [self addSubview:self.mainScrollView];
        [self addSubview:self.barScrollView];
        [self addSubview:self.sendEmojiButton];
        self.configures = configures;
    }
    return self;
}

- (void)buildUI {
    NSArray<LRSMemePackageConfigure *> *configures = self.configures;
    UIView *mainContentView = [[UIView alloc] init];
    [self.mainScrollView addSubview:mainContentView];
    [mainContentView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self.mainScrollView);
    }];
    [self.mainScrollView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.top.right.equalTo(self);
        make.bottom.equalTo(self.barScrollView.mas_top);
    }];
    UIView *barContentView = [[UIView alloc] init];
    [self.barScrollView addSubview:barContentView];
    [barContentView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self.barScrollView);
    }];
    [self.barScrollView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(self);
        make.right.equalTo(self.sendEmojiButton.mas_left);
        make.height.equalTo(self.sendEmojiButton);
        make.centerY.equalTo(self.sendEmojiButton);
    }];

    for (NSInteger index = 0; index < configures.count; ++ index) {
        LRSMemePackageConfigure *configure = configures[index];
        LRSMemePackageControlView *emojiView = [[LRSMemePackageControlView alloc] initWithEmojis:configure.emojis row:configure.row column:configure.column];
        [mainContentView addSubview:emojiView];
        emojiView.itemHandler = self.itemHandler;
        emojiView.backspaceHandler = self.backspaceHandler;
        [emojiView buildUIWithTotalHeight:LRSMemePackagesView.memeBoardHeight];
        [emojiView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.mas_equalTo([self screenWidth] * index);
            make.top.bottom.equalTo(mainContentView);
            make.width.mas_equalTo([self screenWidth]);
            if (index == configures.count - 1) {
                make.right.equalTo(mainContentView);
            }
            make.height.mas_equalTo(LRSMemePackagesView.memeBoardHeight);
        }];

        UIButton *coverButton = [UIButton buttonWithType:(UIButtonTypeCustom)];
        [coverButton setImage:[LRSMessageToolBarHelper imageNamed:configure.coverImageName] forState:(UIControlStateNormal)];
        coverButton.backgroundColor = [LRSMessageToolBarHelper colorNamed:@"Color_244"];
        [barContentView addSubview:coverButton];
        [coverButton mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.mas_equalTo(60 * index);
            make.top.bottom.equalTo(barContentView);
            make.width.mas_equalTo(60);
            if (index == configures.count - 1) {
                make.right.equalTo(barContentView);
            }
            make.height.mas_equalTo(LRSMemePackagesView.actionButtonHeight);
        }];
    }

    [self.sendEmojiButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.right.equalTo(self);
        make.bottom.mas_equalTo(-LRSMessageToolBarHelper.safeAreaHeight);
        make.width.mas_equalTo(60);
        make.height.mas_equalTo(LRSMemePackagesView.actionButtonHeight);
    }];
}

+ (CGFloat)boardHeight {
    return [self memeBoardHeight] + [LRSMessageToolBarHelper safeAreaHeight] + [self actionButtonHeight];
}

+ (CGFloat)memeBoardHeight {
    return 213;
}

+ (CGFloat)actionButtonHeight {
    return 40;
}

- (CGFloat)screenWidth {
    return [LRSMessageToolBarHelper screenWidth];
}

#pragma mark - 表情代理方法

- (UIButton *)sendEmojiButton {
    if (!_sendEmojiButton) {
        _sendEmojiButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [_sendEmojiButton setBackgroundImage:[LRSMessageToolBarHelper imageNamed:@"information_send"] forState:UIControlStateNormal];
        [_sendEmojiButton setTitle:@"发送" forState:UIControlStateNormal];
        [_sendEmojiButton setTitleColor:[LRSMessageToolBarHelper colorNamed:@"Color_30"] forState:UIControlStateNormal];
        _sendEmojiButton.titleLabel.font = [UIFont systemFontOfSize:15.0 weight:UIFontWeightSemibold];
        _sendEmojiButton.titleEdgeInsets = UIEdgeInsetsMake(0, 10, 0, 0);
        [[_sendEmojiButton imageView] setContentMode:UIViewContentModeScaleAspectFit];
    }
    return _sendEmojiButton;
}

- (UIScrollView *)mainScrollView {
    if (!_mainScrollView) {
        UIScrollView *scrollView = [[UIScrollView alloc] init];
        scrollView.pagingEnabled = YES;
        scrollView.showsHorizontalScrollIndicator = NO;
        scrollView.showsVerticalScrollIndicator = NO;
        scrollView.delegate = self;
        _mainScrollView = scrollView;
    }
    return _mainScrollView;
}

- (UIScrollView *)barScrollView {
    if (!_barScrollView) {
        UIScrollView *scrollView = [[UIScrollView alloc] init];
        scrollView.showsHorizontalScrollIndicator = NO;
        scrollView.showsVerticalScrollIndicator = NO;
        scrollView.delegate = self;
        _barScrollView = scrollView;
    }
    return _barScrollView;
}
@end

File: LRSMessageToolBar/Classes/Emoji/LRSMemePackagesView.swift
//
//  LRSMemePackagesView.swift
//  LRSMessageToolBarComponents
//
//  Created by sama 刘 on 2021/11/17.
//

import UIKit

class LRSMemePackagesView: UIView {

    public typealias ItemHandler = (LRSMemePackagesView) -> ()

    public var boardHeight: CGFloat {
        return memeBoardHeight + LRSMessageToolBarHelper.safeAreaHeight + actionButtonHeight
    }

    public var memeBoardHeight: CGFloat {
        return 213
    }

    public var actionButtonHeight: CGFloat {
        return 40
    }


    lazy var mainScrollView: UIScrollView = UIScrollView().then {
        $0.isPagingEnabled = true
        $0.showsHorizontalScrollIndicator = false;
        $0.showsVerticalScrollIndicator = false
    }

    lazy var barScrollView: UIScrollView = UIScrollView().then{
        $0.showsHorizontalScrollIndicator = false;
        $0.showsVerticalScrollIndicator = false
    }

    lazy var sendEmojiButton: UIButton = UIButton(type: .custom).then {
        $0.setBackgroundImage(.image(named: "information_send"), for: .normal)
        $0.setTitle("发送", for: .normal)
        $0.setTitleColor(UIColor.color(named: "Color_30"), for: .normal)
        $0.titleLabel?.font = .systemFont(ofSize: 15, weight: .bold)
        $0.titleEdgeInsets = UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10)
        $0.imageView?.contentMode = .scaleAspectFit
        $0.addTarget(self, action: #selector(sendOut), for: .touchUpInside)
    }

    let itemHandler: LRSMemeSinglePage.ItemHandler
    let deleteHandler: LRSMemeSinglePage.ItemHandler
    let confirmHandler: ItemHandler

    init( itemHandler: @escaping LRSMemeSinglePage.ItemHandler,
         deleteHandler: @escaping LRSMemeSinglePage.ItemHandler,
         confirmHandler: @escaping ItemHandler) {
        self.itemHandler = itemHandler
        self.deleteHandler = deleteHandler
        self.confirmHandler = confirmHandler
        super.init(frame: .zero)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    @objc private func sendOut() {
        confirmHandler(self)
    }

    func buildUI(configures: [LRSMemePackageConfigure]) {
        addSubview(mainScrollView)
        addSubview(barScrollView)
        addSubview(sendEmojiButton)

        let mainContentView = UIView().then {
            mainScrollView.addSubview($0)
            $0.snp.makeConstraints { make in
                make.edges.equalToSuperview()
            }
        }
        mainScrollView.snp.makeConstraints { make in
            make.left.top.right.equalToSuperview()
            make.bottom.equalTo(barScrollView.snp.top)
        }

        let barContentView = UIView().then {
            barScrollView.addSubview($0)
            $0.snp.makeConstraints { make in
                make.edges.equalToSuperview()
            }
        }
        barScrollView.snp.makeConstraints { make in
            make.left.equalToSuperview()
            make.right.equalTo(sendEmojiButton.snp.left)
            make.centerY.equalTo(sendEmojiButton)
            make.height.equalTo(sendEmojiButton)
        }

        sendEmojiButton.snp.makeConstraints { make in
            make.right.equalToSuperview()
            make.bottom.equalTo(-LRSMessageToolBarHelper.safeAreaHeight)
            make.width.equalTo(60)
            make.height.equalTo(actionButtonHeight)
        }

        configures.enumerated().forEach { (offset, element) in
            let size = LRSMessageToolBarHelper.sizeOfEmoji(key: element.key)
            let emojiView = LRSMemePackageControlView(
                itemHandler: itemHandler,
                deleteHandler: deleteHandler,
                uiConfigure: LRSMemePackageControlView.Configure(
                    buttonSize: size,
                    rows: element.row,
                    columns: element.column
                ),
                emojis: element.emojis
            )
            mainContentView.addSubview(emojiView)
            emojiView.buildUIWithTotalHeight(memePackageBoardHeight: memeBoardHeight)
            emojiView.snp.makeConstraints { make in
                make.left.equalTo(UIScreen.main.width * CGFloat(offset))
                make.top.bottom.equalToSuperview()
                make.width.equalTo(UIScreen.main.width)
                if (offset == configures.count - 1) {
                    make.right.equalToSuperview()
                }
                make.height.equalTo(memeBoardHeight);
            }

            let coverButton = UIButton(type: .custom)
            coverButton.setImage(.image(named: element.coverImageName), for: .normal)
            coverButton.backgroundColor = .color(named: "Color_244")
            barContentView.addSubview(coverButton)
            coverButton.snp.makeConstraints { make in
                make.left.equalTo(actionButtonHeight * CGFloat(offset));
                make.top.bottom.equalToSuperview()
                make.height.width.equalTo(actionButtonHeight);
                if (offset == configures.count - 1) {
                    make.right.equalToSuperview()
                }
            }
        }
    }

}
